Mathlib/Tactic/Elementwise.lean

```diff
@@ -0,0 +1,213 @@
+If `simp` is added first, the generated lemma will also have the `simp` attribute.
+ -/
+syntax (name := elementwise) "elementwise" ("(" &"attr" ":=" Parser.Term.attrInstance,* ")")? : attr
+
+initialize registerBuiltinAttribute {
+  name := `elementwise
+  descr := ""
+  applicationTime := .afterCompilation
+  add := fun src ref kind => match ref with
+  | `(attr| elementwise $[(attr := $stx?,*)]?) => MetaM.run' do
+    let tgt := match src with
+      | Name.str n s => Name.mkStr n $ s ++ "_apply"
+      | x => x
+    if (kind != AttributeKind.global) then
+      throwError "`elementwise` can only be used as a global attribute"
+    addDeclarationRanges tgt
+      { range := ← getDeclarationRange (← getRef)
+        selectionRange := ← getDeclarationRange ref }
+    let info ← getConstInfo src
+    let (newValue, level?) ← elementwiseExpr info.type info.value!
+    let mut levels := info.levelParams
+    if let some level := level? then
+      let w := mkUnusedName levels `w
+      unless ← isLevelDefEq level (mkLevelParam w) do
+        throwError "Could not create level parameter for ConcreteCategory instance"
+      levels := w :: levels
+    let newValue ← instantiateMVars newValue
+    let newType ← instantiateMVars (← inferType newValue)
+    match info with
+    | ConstantInfo.thmInfo info =>
+      addAndCompile <| .thmDecl
+        { info with levelParams := levels, type := newType, name := tgt, value := newValue }
+    | ConstantInfo.defnInfo info =>
+      -- It looks a bit weird that we use `.thmDecl` here too,
+      -- but apparently structure fields are created using `def`
+      -- even with they are propositional. If `elementwise` worked, it was a `Prop` anyway.
+      addAndCompile <| .thmDecl
+        { info with levelParams := levels, type := newType, name := tgt, value := newValue }
+    | _ => throwError "Constant {src} is not a theorem or definition."
+    if isProtected (← getEnv) src then
+      setEnv $ addProtected (← getEnv) tgt
+    let stx := match stx? with | some stx => stx | none => #[]
+    _ ← Term.TermElabM.run' <| ToAdditive.applyAttributes ref stx `elementwise src tgt
+  | _ => throwUnsupportedSyntax }
+
+/--
+`elementwise_of% h`, where `h` is
+an equation `f = g` between morphisms `X ⟶ Y` in a category (possibly after a `∀` binder),
+produces the equation `∀ (x : X), f x = g x`, but with compositions fully right associated and
+identities removed. The category must already have a `ConcreteCategory` instance.
+
+The type and proof of `%elementwise_of h` is generated by `Tactic.Elementwise.elementwiseExpr`
+which makes `%elementwise_of` meta-programming adjacent. It is not called as a tactic but as
+an expression. The goal is to avoid creating assumptions that are dismissed after one use:
+```lean
+example (M N K : Mon) (f : M ⟶ N) (g : N ⟶ K) (h : M ⟶ K) (w : f ≫ g = h) (m : M) :
+    g (f m) = h m := by rw [%elementwise_of w]
+```
```

Comment:
```suggestion
The type and proof of `elementwise_of% h` is generated by `Tactic.Elementwise.elementwiseExpr`
which makes `elementwise_of%` meta-programming adjacent. It is not called as a tactic but as
an expression. The goal is to avoid creating assumptions that are dismissed after one use:
```lean
example (M N K : Mon) (f : M ⟶ N) (g : N ⟶ K) (h : M ⟶ K) (w : f ≫ g = h) (m : M) :
    g (f m) = h m := by rw [elementwise_of% w]
```
```

That said, I think we could just drop this second paragraph. Term-elaborators are just a regular thing in the Lean 4 world, and we don't need to say things like "meta-programming adjacent".

Reply:
Thanks, I kept writing `%elementwise_of` in tests too. 

Improving this documentation got me thinking -- there's no real reason to prevent `elementwise_of%` from generating lemmas that need a fresh level metavariable (after all, every rewrite lemma is instantiated with fresh level metavariables), so I've removed that restriction. This should mean that you can either add `@[elementwise]` to a pre-existing lemma or rewrite using `elementwise_of%`, pretty much interchangeably. (I didn't copy it from the original `elementwise_of`; this restriction was a new feature that I didn't think it through.)